Added insertion order support.

redboltz opened the pull request:
Added cmd type check on put. If cmd type is the same before and after
update, update is aboted.

In the MQTT communication, update is only required from 'publish' to 'pubrel'.

redboltz committed:
Added insertion order support.

Added cmd type check on put. If cmd type is the same before and after
update, update is aboted.

In the MQTT communication, update is only required from 'publish' to 'pubrel'.

mcollina commented on mqtt-level-store.js:
```diff
@@ -19,38 +19,112 @@ function Store (options) {
+
 Store.prototype.put = function (packet, cb) {
-  this._level.put('' + packet.messageId, packet, this._levelOpts, cb)
+  var date = zeroPadding(Date.now())
```
even though is slower, I would prefer to use `new Date().toISOString()`. That is lexicographically safe.

redboltz replied:
Ok, I will replace.

mcollina commented on mqtt-level-store.js:
```diff
@@ -19,38 +19,112 @@ function Store (options) {
-  this._level.put('' + packet.messageId, packet, this._levelOpts, cb)
+  var date = zeroPadding(Date.now())
+  if (this._prevDate === date) {
+    ++this._sameDateCount
```
why is this needed?

redboltz replied:
`Date.now()` resolution is millisecond. When user does continuous publish, the `Date.now()` could be the same value as the previous one. And at that time messageId is back to 1, order would be broken. So I introduced the counter.

mcollina replied:
Can you please initialize this variable in the constructor?

redboltz replied:
Yes. I will do as follows.

```javascript
function Store (options) {
  if (!(this instanceof Store)) {
    return new Store(options)
  }

  if (!options.level) {
    throw new Error('missing level')
  }

  this._level = options.level
  this._levelOpts = {
    valueEncoding: msgpack()
  }
  this._sameDateCount = 0
  this._prevDate = null
}
```

mcollina commented on mqtt-level-store.js:
```diff
@@ -19,38 +19,112 @@ function Store (options) {
+            if (packet.cmd === _packet.cmd) {
+              cb()
+            } else {
+              that._level.put('packet-by-date~' + _date + '~' + packet.messageId, packet, that._levelOpts,
```
can you just pass `cb` down?

redboltz replied:
I guess you mean line 65 and 66. I will fix it.

redboltz committed:
Adjusted timeout value to meet travis-ci environment.

mcollina commented on mqtt-level-store.js:
```diff
@@ -19,38 +19,112 @@ function Store (options) {
+        } else {
+          cb(err)
+        }
+      } else {
```
can you just return early from the previous block instead of adding this else?

redboltz replied:
Do you mean as follows?

```cpp
      if (err) {
        if (!err.notFound) return cb(err)
        var cmd = [
          {type: 'put', key: 'packets~' + packet.messageId, value: date},
          {type: 'put', key: 'packet-by-date~' + date + '~' + packet.messageId, value: packet}
        ]
        that._level.batch(cmd, that._levelOpts, cb)
      } else {
```

Or

```cpp
        if (!err.notFound) return cb(err)
```

is

```cpp
        if (!err.notFound) {
          cb(err)
          return this
        }
```

?

Or something else?

redboltz replied:
I understand the first one is appropriate. Similar to https://github.com/mcollina/mqtt-level-store/pull/11#discussion_r212822705

mcollina commented on mqtt-level-store.js:
```diff
@@ -19,38 +19,112 @@ function Store (options) {
+            if (err) return cb(err)
+            if (packet.cmd === _packet.cmd) {
+              cb()
+            } else {
```
can you return early from the previous block instead?

redboltz replied:
Do you mean this?

```javascript
            if (packet.cmd !== _packet.cmd) return cb()
              that._level.put('packet-by-date~' + _date + '~' + packet.messageId, packet, that._levelOpts, cb)
            }
```

mcollina replied:
yes

mcollina commented on mqtt-level-store.js:
```diff
@@ -19,38 +19,112 @@ function Store (options) {
+        that._levelOpts,
+        function (err, packet) {
+          cb(err, packet)
+        })
```
can you just pass `cb` to `_level.get()`?

redboltz replied:
Yes, I will fix it.

mcollina commented on mqtt-level-store.js:
```diff
@@ -19,38 +19,112 @@ function Store (options) {
+    this._levelOpts,
+    function (err, date) {
+      if (err) return cb(err)
+      that._level.get(
```
I think this  `get` is not really needed.

redboltz replied:
It is for getting date. Then get data using the got date as a part of the key.

mcollina replied:
the date is the output of the previous get, right?

redboltz replied:
Yes, the second parameter of the callback.

redboltz replied:
```js
  this._level.get(
    'packets~' + packet.messageId,  // get date from messageId
    this._levelOpts,
    function (err, date) {
      if (err) return cb(err)
      that._level.get( 
        'packet-by-date~' + date + '~' + packet.messageId, // use date as a part of the key for value
        that._levelOpts,
        function (err, _packet) {
```

redboltz committed:
Refactoring code.

Replaced `Date.now()` with `new Date().toISOString()`
Initialized _prevDate and _sameDateCount in the constructor.
Refined if/else logic.

redboltz committed:
Added `manager.close()` to afterEach on `mqtt.connect flow`.

Timeout value back to 200.

redboltz commented:
I fixed all code that you pointed out. Could you check it?

Some of `'should resend messages by published order'` on travis-ci reported 2000ms timeout.
It doesn't happen on my local environment.
Could you restart failed test. Or should I do something?

redboltz commented:
I tested the PR with debug print on my travis-ci environment.

https://travis-ci.org/redboltz/mqtt-level-store/jobs/420888797#L715

```js
  beforeEach(function (done) {
    console.log('new server')
    server = new net.Server()
    console.log('server listen')
    server.listen(8883, done)

    console.log('server before connect')
    server.on('connection', function (stream) {
      console.log('server after connect')
      var client = Connection(stream)

      client.on('connect', function () {
        console.log('client on connect')
        client.connack({returnCode: 0})
      })

      console.log('emit client')
      server.emit('client', client)
    })

    console.log('create level store')
    manager = mqttLevelStore({ level: level() })
  })

  afterEach(function (done) {
    console.log('after each')
    manager.close(function () {
      console.log('manager closed')
      server.close(function () {  // *** block here ***
        console.log('server closed')
        done()
      })
    })
  })
```

I noticed that server.close callback sometimes not called on travis-ci. Maybe there is a better way.
I started read MQTT.js's test code. If you know something about that, please give me your advise.

redboltz commented:
During try and error, I commented out the test `'should resend messages'` https://github.com/mcollina/mqtt-level-store/blob/master/test.js#L64 , the test `'should resend messages by published order'` passed.
It indicates that there is some problem to do multiple tests in the test suite `'mqtt.connect flow'`.

I think that we have two choice.

1. Remove `'should resend messages'`. The test is to check resend one message. `'should resend messages by published order'` is to check resend three messages. I think that the latter covers the former. It is the easiest way for me.

2. Introduce a proper way to do multiple tests in the test suite `'mqtt.connect flow'` on travis-ci. I think that is the best way. But I don't know how to do it. The current master seems to support only one test.

redboltz commented:
Here is a diff of approach 1 branch and master:

[diff](https://github.com/mcollina/mqtt-level-store/compare/master...redboltz:replace_test#diff-1dd241c4cd3fd1dd89c570cee98b79ddL64)

And travis-ci result(stable success):
https://travis-ci.org/redboltz/mqtt-level-store/builds/420910228

mcollina reviewed (approved):
LGTM

mcollina commented:
CI is a bit flaky, can you improve the stability of the tests?
Also, can you add Node 10?

redboltz commented:
Do you mean I can use approach 1? It’s very stable. And I will add node 10.

mcollina commented:
Yeah, you can do 1.

redboltz committed:
Added node.js 9 to travis-ci.

Unified tests in mqtt.connect flow to increase travis-ci stability.

redboltz commented:
I bumped a [compile error](https://travis-ci.org/redboltz/mqtt-level-store/jobs/421076258#L667)  when compiling leveldown on nodejs 10.

I googled the error and got many [error reports](https://www.google.co.jp/search?q=leveldb+..%2F..%2Fnan%2Fnan_maybe_43_inl.h%3A112%3A15%3A+error%3A+%E2%80%98class+v8%3A%3AObject%E2%80%99+has+no+member+named+%E2%80%98ForceSet%E2%80%99&oq=leveldb+..%2F..%2Fnan%2Fnan_maybe_43_inl.h%3A112%3A15%3A+error%3A+%E2%80%98class+v8%3A%3AObject%E2%80%99+has+no+member+named+%E2%80%98ForceSet%E2%80%99&aqs=chrome..69i57.1829j0j1&sourceid=chrome&ie=UTF-8).

I couldn't find a way to solve it. But it is not directly related to mqtt-level-store. So I added nodejs 9 instead of 10. I think that it is acceptable for now.

mcollina commented:
That's perfect, thanks!
